src/field_theory/splitting_field.lean

```diff
@@ -109,46 +109,392 @@ namespace splitting_field_aux
+section lift_instances
 
-instance inhabited {n : ℕ} {f : K[X]} :
-  inhabited (splitting_field_aux n f) := ⟨37⟩
+/-! ### Instances on `splitting_field_aux`
 
-/-
-Note that the recursive nature of this definition and `splitting_field_aux.field` creates
-non-definitionally-equal diamonds in the `ℕ`- and `ℤ`- actions.
-```lean
-example (n : ℕ) {K : Type u} [field K] {f : K[X]} (hfn : f.nat_degree = n) :
-    (add_comm_monoid.nat_module : module ℕ (splitting_field_aux n f hfn)) =
-  @algebra.to_module _ _ _ _ (splitting_field_aux.algebra n _ hfn) :=
-rfl  -- fails
-```
-It's not immediately clear whether this _can_ be fixed; the failure is much the same as the reason
-that the following fails:
-```lean
-def cases_twice {α} (a₀ aₙ : α) : ℕ → α × α
-| 0 := (a₀, a₀)
-| (n + 1) := (aₙ, aₙ)
-
-example (x : ℕ) {α} (a₀ aₙ : α) : (cases_twice a₀ aₙ x).1 = (cases_twice a₀ aₙ x).2 := rfl  -- fails
-```
-We don't really care at this point because this is an implementation detail (which is why this is
-not a docstring), but we do in `splitting_field.algebra'` below. -/
-instance algebra (n : ℕ) : Π (R : Type*) {K : Type u} [comm_semiring R] [field K],
-  by exactI Π [algebra R K] {f : K[X]},
-    algebra R (splitting_field_aux n f) :=
-nat.rec_on n (λ R K _ _ _ _, by exactI ‹algebra R K›) $
-         λ n ih R K _ _ _ f, by exactI ih R
+In order to avoid diamond issues, we have to be careful to define each field of
+an instance of `splitting_field_aux` by recursion, rather than defining the
+whole structure by recursion at once.
+
+We do this in three steps: first define an `add_comm_group` instance to put a `distrib_smul`
+on (which is what is needed to push scalar multiplication through the recursive step),
+then a `comm_ring` instance, and finally a `field` instance.
+-/
+
+/-- Splitting fields have a zero. -/
+protected def zero (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]},
+  splitting_field_aux n f :=
+nat.rec_on n (λ K fK f, by exactI @has_zero.zero K _) (λ n ih K fK f, ih)
+
+/-- Splitting fields have an addition. -/
+protected def add (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]},
+  splitting_field_aux n f → splitting_field_aux n f → splitting_field_aux n f :=
+nat.rec_on n (λ K fK f, by exactI @has_add.add K _) (λ n ih K fK f, ih)
+
+/-- `K` adjoined `n` roots of `f : K[X]` has addition and zero.
+
+We stop via this instance on our path to the `field` instance
+in order to define the `distrib_smul` instance below; a `distrib_smul`
+is required to lift scalar multiplications on `K` to a splitting field of `f : K[X]`.
+-/
+instance add_zero_class (n : ℕ) {K : Type u} [field K] {f : K[X]} :
+  add_zero_class (splitting_field_aux n f) :=
+begin
+  refine_struct
+    { add := splitting_field_aux.add n,
+      zero := splitting_field_aux.zero n },
+  all_goals { unfreezingI { induction n with n ih generalizing K } },
+  -- The `succ` cases follow by induction, handle them first.
+  iterate 2 { rotate, exact ih },
```

Comment:
This seems super fragile; can we refer to goals by name instead?

Reply:
it's fine enough here but on the `comm_ring` section you end up getting 12 goals that you have to prove the same way

Reply:
In theory we should be able to use the `case` tactic here. Unfortunately, `unfreezingI` discards case names (leanprover/lean#804). Without that PR, we can use
```lean
propagate_tags { unfreeze_local_instances }, tac, all_goals { freeze_local_instances }
```
I'll have a go at that and see what happens.